### PR TITLE
Changed simulators fallback to `decompose_once` and removed ancilla support from `DensityMatrixSimulator`

### PR DESCRIPTION
**Pull request by tanujkhattar**
See https://github.com/quantumlib/Cirq/issues/6081#issuecomment-1582107970 for a detailed description. Tl;Dr is as follows:

1) Fixed bugs in `strat_act_on_from_apply_decompose` and `SimulationState` which were introduced as part of https://github.com/quantumlib/Cirq/pull/6108
2) Laid foundations for choosing a custom qubit allocation strategy when simulating circuits containing operations that can allocate new ancillas as part of their decompose protocol.

### CODE REVIEW

**Comment by NoureldinYosri** [comment]
The title of the PR doesn't really match the content of the PR, how about "Changed simulators' fallback to `decompose_once` and removed ancilla support from `DensityMatrixSimulator`". 

does having DensityMatrixSimulator support ancillas block cirq-ft ?

---

nit: this PR is 2 in 1: the first a partial reversal of https://github.com/quantumlib/Cirq/pull/6108 to remove the ancilla support from `DensityMatrixSimulator` and the second one contains the changes to `strat_act_on_from_apply_decompose`. so splitting it might be a good idea.

**Comment by tanujkhattar** [comment]
Updated the title. 

> does having DensityMatrixSimulator support ancillas block cirq-ft ?

No.

> so splitting it might be a good idea

Hmm, I'll let people take a look at it as is for a bit and I can split if it's getting hard to review.

**Comment by NoureldinYosri** [comment]
LGTM. pending review from @senecameeks

**Review by senecameeks — COMMENTED** [review]
Reviewed, looks good!

**Review by NoureldinYosri — APPROVED** [review]
please update the docstring of `add_qubits` in `simulation_state` before you merge

File: cirq-core/cirq/sim/simulation_state.py
Comment (tanujkhattar) [reply cirq-core/cirq/sim/simulation_state.py]: Good catch. Also updated `remove_qubits` docstring since it had some typos.

**Pull request merged**
